server/gameserver.js

```diff
@@ -82,6 +82,37 @@ async function ParseModPDiffs( request )
 	return modInfo
 }
 
+// Uses 3rd-party service to get continent of gameserver based on GeoIP
+async function getGeoIp( request )
+{
+	let geoIpResonse
+	try
+	{
+		geoIpResonse = await asyncHttp.request( {
+			method: "GET",
+			host: "ip-api.com",
+			path: "/json/" + request.ip + "?fields=status,continent"
+		} )
+	}
+	catch
+	{
+		// Something failed, return empty string
+		return ""
+	}
+
+	let json_response = JSON.parse( geoIpResonse.toString() )
+
+	if ( json_response.status === "success" )
+	{
+		return json_response.continent
+	}
+	else
```

Comment:
We should not treat all errors the same way; if rate limit has been reached (status === 429), I think we should log it, and prevent any GeoIP requests until limit is reset, to avoid master server's IP being banned.

Reply:
100% agree.

Reply:
Okay so I did some testing with just spamming the API from my local machine with a Python script and instead of giving back a 429 or similar error, the API just takes forever to respond. That's kinda the only result I didn't wanna see :c

Reply:
I never got a 429 in testing, only 200s even after slowdown.

Reply:
Did the last request contain some additional information in headers (`X-Rl`, `X-Ttl`)?
I'll check tomorrow if I can reproduce this.

Reply:
Yeah I get slow-down before ` X-Rl`  even hits `0` (starts around `30`)

Reply:
On my quest to find alternatives I've been hammering an endpoint of https://geoiplookup.io/api for >1000 requests in 5minutes without slowdowns, so I might switch to that.

Reply:
Okay, I have no idea how to extract status code and headers from the custom `asyncHTTP.request()` function we have ._.

So any help would be appreciated <3

Reply:
> Okay so I did some testing with just spamming the API from my local machine with a Python script and instead of giving back a 429 or similar error, the API just takes forever to respond. That's kinda the only result I didn't wanna see :c

I can confirm API does not respond after reaching rate limit. This is not a standard behaviour...

Reply:
> On my quest to find alternatives I've been hammering an endpoint of https://geoiplookup.io/api for >1000 requests in 5minutes without slowdowns, so I might switch to that.

Are you sure about that rate?

![image](https://user-images.githubusercontent.com/11993538/183604090-04631b95-343c-4ea4-816c-5c39b1b66f06.png)

Reply:
> Okay, I have no idea how to extract status code and headers from the custom `asyncHTTP.request()` function we have ._.
> 
> So any help would be appreciated <3

It seems that the custom async package just throws requests not having a status between 200 and 300:
https://github.com/R2Northstar/NorthstarMasterServer/blob/geoip-region-detection/shared/asynchttp.js#L19

I don't even know why this package exists in the first place.

Reply:
> > On my quest to find alternatives I've been hammering an endpoint of https://geoiplookup.io/api for >1000 requests in 5minutes without slowdowns, so I might switch to that.
> 
> Are you sure about that rate?
> 
> ![image](https://user-images.githubusercontent.com/11993538/183604090-04631b95-343c-4ea4-816c-5c39b1b66f06.png)

Yeah, I saw that hence I did the testing and was kinda surprised with the results ^^

Reply:
Cool!
Tell me if you need help with the `request` implementation.

Reply:
> Tell me if you need help with the `request` implementation.

Given my very limited JavaScript experience, I wouldn't mind :P

Should probably be its own PR and I'll merge it into this one from there ^^

Reply:
> You might want to add a setting to the `request` method, telling it to return status and headers along response body, returning something like:
> 
> ```json
> {
>     "status": 429,
>     "headers": { ... },
>     "body": "Too much requests, try again later."
> }
> ```

Aight, so we have this now when #86 gets merged. ^^